pkg/hydrator/batch.go

```diff
@@ -0,0 +1,155 @@
+package hydrator
+
+import (
+	"context"
+
+	"github.com/asecurityteam/nexpose-vuln-hydrator/pkg/domain"
+)
+
+// BatchSolutionFetcher represents an interface for concurrently fetching solutions
+type BatchSolutionFetcher interface {
+	BatchFetchSolution(ctx context.Context, solutionIDs []string) ([]string, error)
+}
+
+type batchSolutionFetcher struct {
+	SolutionFetcher SolutionFetcher
+}
+
+func (b *batchSolutionFetcher) BatchFetchSolution(ctx context.Context, solutionIDs []string) ([]string, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	solutionsChan := make(chan string, len(solutionIDs))
+	errorChan := make(chan error, 1)
+
+	for _, solutionID := range solutionIDs {
+		go func(id string) {
+			solution, err := b.SolutionFetcher.FetchSolution(ctx, id)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+			solutionsChan <- solution
+		}(solutionID)
+	}
+
+	solutions := make([]string, len(solutionIDs))
+	for i := range solutionIDs {
+		select {
+		case solution := <-solutionsChan:
+			solutions[i] = solution
```

Comment:
task: Using this pattern of adding elements to the slice will make it harder to later support partial errors if we decide that's what we need. A pattern that works for all cases is:

```golang
// Generate a slice with an underlying array that is already big enough
// to contain the data we need but is set to size of zero.
solutions := make([]string, 0, len(solutionIDs))
for range solutionIDs {
  select {
  case solution := <-solutionsChan:
      solutions = append(solutions, solution)
  }
}
```

This pattern provides the same optimization of pre-allocating the slice but also comes with the benefit of handling cases where errors _don't_ exit the function. Even if we don't need this behavior for this specific instance, I recommend using this pattern everywhere we populate a pre-allocated size slice.